feat: support disag serving in GAIE [DEP-659]

atchernych opened the pull request:
#### Overview:

[DEP-659](https://linear.app/nvidia/issue/DEP-659/support-disaggregated-serving-in-gaie-integration) 

For GAIE stage 1 the response is expected to be in the format:
```
{
  "id": "chatcmpl-xxx",
  "object": "chat.completion.chunk",
  "choices": [...],
  "nvext": {
    "worker_id": {
      "prefill_worker_id": 12345,
      "decode_worker_id": 67890
    },
    "token_ids": [101, 2023, 2003, 1037, 3231, 6251, 102]
  }
}

For agg prefill_worker_id = decode_worker_id

For GAIE stage 2, GAIE will send the request in the format:
Disag:

{
  "model": "my-model",
  "messages": [...],
  "nvext": {
    "prefill_worker_id": 12345,
    "decode_worker_id": 67890,
    "token_data": [101, 2023, 2003, 1037, 3231, 6251, 102]
  }
}

Agg:


{
  "model": "my-model",
  "messages": [...],
  "nvext": {
    "backend_instance_id": 54321,
    "token_data": [101, 2023, 2003, 1037, 3231, 6251, 102]
  }
}


```
<!-- Describe your pull request here. Please read the text below the line, and make sure you follow the checklist.-->

#### Details:

<!-- Describe the changes made in this PR. -->

#### Where should the reviewer start?

<!-- call out specific files that should be looked at closely -->

#### Related Issues: (use one of the action keywords Closes / Fixes / Resolves / Relates to)

- closes GitHub issue: #xxx


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->

## Summary by CodeRabbit

## New Features

* **Disaggregated Serving**: Added support for staged inference execution, enabling prefill and decode operations to be routed to different workers independently.
* **Query-Only Workflow**: Introduced capability to query worker assignments for planned inference without executing the full inference pass.
* **Enhanced Response Data**: Responses now include query stage completion status and detailed worker assignment information for improved request tracking and routing visibility.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

PeaBrane commented:
@atchernych thanks for your contribution! Can you please take a look at my suggestion

## Suggestion: Simplify GAIE Stage 1 with a state-machine approach

The current implementation adds significant complexity to `PrefillRouter` by having it explicitly coordinate both prefill and decode worker selection. This could be simplified by using `query_instance_id` as a state marker that flows through the pipeline.

### Current approach
- `PrefillRouter` intercepts `query_instance_id`, manually queries both routers, and constructs the combined response
- Adds `query_only` parameter, `PrefillCallResult::WorkerIdOnly` variant, and ~90 lines of coordination logic
- Duplicates worker ID extraction logic in multiple places

### Suggested approach
Use `query_instance_id` as a state machine marker:

```rust
// State transitions:
// "disagg" → "prefill" → "decode" → response
// "agg"    → response (single worker for both)

// PrefillRouter behavior:
if query_instance_id == "disagg":
    req.query_instance_id = "prefill"
    prefill_response = prefill_router.generate(req)
    prefill_worker_id = extract_worker_id(prefill_response)
    
    req.query_instance_id = "decode"
    req.prefill_worker_id = prefill_worker_id  # attach for decode router
    return next.generate(req)  # decode router handles the rest

// KvRouter behavior:
match query_instance_id:
    case "prefill":
        return {"worker_id": {"prefill_worker_id": selected_id}}
    case "decode":
        return {"worker_id": {"prefill_worker_id": req.prefill_worker_id, "decode_worker_id": selected_id}}
    case "agg":
        return {"worker_id": {"prefill_worker_id": selected_id, "decode_worker_id": selected_id}}
```

### Benefits
1. Each component only handles its own transformation
2. No special coordination logic in `PrefillRouter` - it just transforms the marker and passes through
3. Removes `query_only` param, `PrefillCallResult` enum, and `select_prefill_decode_workers` (~90 lines)
4. Aggregated mode falls out naturally (`"agg"` sets both IDs to the same worker)
5. Existing pipeline structure handles the flow - no manual querying of decode router

atchernych committed:
Support disag in GAIE

Signed-off-by: Anna Tchernych <atchernych@nvidia.com>

atchernych committed:
rm disagg and agg enums

Signed-off-by: Anna Tchernych <atchernych@nvidia.com>

atchernych committed:
cleanup lib.rs

Signed-off-by: Anna Tchernych <atchernych@nvidia.com>

atchernych committed:
style fix

Signed-off-by: Anna Tchernych <atchernych@nvidia.com>

atchernych committed:
comment addressed

Signed-off-by: Anna Tchernych <atchernych@nvidia.com>

atchernych committed:
use query_instance_id as a state machine

Signed-off-by: Anna Tchernych <atchernych@nvidia.com>

atchernych committed:
unnify decode flow

Signed-off-by: Anna Tchernych <atchernych@nvidia.com>

atchernych committed:
rm GAIE prefill error handler case

Signed-off-by: Anna Tchernych <atchernych@nvidia.com>

atchernych committed:
unify prefill response

Signed-off-by: Anna Tchernych <atchernych@nvidia.com>

atchernych committed:
fix the format

Signed-off-by: Anna Tchernych <atchernych@nvidia.com>

atchernych committed:
account for agg flow

Signed-off-by: Anna Tchernych <atchernych@nvidia.com>

atchernych committed:
cleanup update_states

Signed-off-by: Anna Tchernych <atchernych@nvidia.com>

atchernych committed:
update lib.rs bidnings

Signed-off-by: Anna Tchernych <atchernych@nvidia.com>

atchernych committed:
move up for fewer changes

Signed-off-by: Anna Tchernych <atchernych@nvidia.com>

PeaBrane commented on lib/llm/src/kv_router.rs:
```diff
@@ -722,13 +764,34 @@ impl AsyncEngine<SingleIn<PreprocessedRequest>, ManyOut<Annotated<LLMEngineOutpu
+                        // Empty value is valid for aggregated mode, not a warning
+                        None
+                    }
+                    Err(e) => {
```
you can inline them as {type_str} {e} etc

atchernych replied:
will do

PeaBrane commented on lib/llm/src/kv_router.rs:
```diff
@@ -752,33 +815,90 @@ impl AsyncEngine<SingleIn<PreprocessedRequest>, ManyOut<Annotated<LLMEngineOutpu
+            };
+
+            // Return as LLMEngineOutput with disaggregated_params (same structure as normal execution)
+            let output = LLMEngineOutput {
```
Not your PR. But let's derive a default for LLMEngineOutput, so we can just use a default here and then set `disaggregated_params`

PeaBrane commented on lib/llm/src/kv_router/prefill_router.rs:
```diff
@@ -272,21 +291,99 @@ impl
         let request_id = context.id().to_string();
         let engine_ctx = context.context();
 
+        // GAIE Stage 1: the presence of the empty query_instance_id signals query-only mode
```
too much code duplication, can we put this GAIE handling logic into a helper function / method

atchernych committed:
fixes per comments

Signed-off-by: Anna Tchernych <atchernych@nvidia.com>

atchernych committed:
cleanup

Signed-off-by: Anna Tchernych <atchernych@nvidia.com>

atchernych committed:
more cleanup

Signed-off-by: Anna Tchernych <atchernych@nvidia.com>

atchernych committed:
rm prefll active check

Signed-off-by: Anna Tchernych <atchernych@nvidia.com>

atchernych committed:
merge main

Signed-off-by: Anna Tchernych <atchernych@nvidia.com>

atchernych committed:
allow for bootstrap opt

Signed-off-by: Anna Tchernych <atchernych@nvidia.com>

atchernych committed:
Update the patch for disag

Signed-off-by: Anna Tchernych <atchernych@nvidia.com>

atchernych committed:
merge main

atchernych committed:
fix: update test to match new query_instance_id response format

- Use 'query_instance_id:' (with colon) for annotation format
- Parse worker_id and token_ids from nvext field instead of SSE events
- Validate prefill_worker_id == decode_worker_id for aggregated mode

Signed-off-by: Anna Tchernych <atchernych@nvidia.com>
